src/build.ts

```diff
@@ -1755,7 +1768,68 @@ class LocalTokenSet extends TokenSet {
       }
     }
 
-    let precTable = this.buildPrecTable(none)
+
+    let groups: TokenGroup[] = []
+    for (let state of states) {
+      if (state.defaultReduce || state.tokenGroup > -1) continue
+      // Find potentially-conflicting terms (in terms) and the things
+      // they conflict with (in conflicts), and raise an error if
+      // there's a token conflict directly in this state.
+      let terms: Term[] = [], incompatible: Term[] = []
+      let skip = skipInfo[this.b.skipRules.indexOf(state.skip)].startTokens
+      for (let term of skip)
+        if (state.actions.some(a => a.term == term))
+          this.b.raise(`Use of token ${term.name} conflicts with skip rule`)
+
+      let stateTerms: Term[] = []
+      for (let i = 0; i < state.actions.length + (skip ? skip.length : 0); i++) {
+        let term = i < state.actions.length ? state.actions[i].term : skip[i - state.actions.length]
+        let orig = this.b.tokenOrigins[term.name]
+        if (orig && orig.spec) term = orig.spec
+        else if (orig && orig.external) continue
+        addToSet(stateTerms, term)
+      }
+      if (stateTerms.length == 0) continue
+
+
+      for (let term of stateTerms) {
+        for (let conflict of conflicts) {
+          let conflicting = conflict.a == term ? conflict.b : conflict.b == term ? conflict.a : null
+          if (!conflicting) continue
+          if (stateTerms.includes(conflicting) && !errors.some(e => e.conflict == conflict)) {
+            let example = conflict.exampleA ? ` (example: ${JSON.stringify(conflict.exampleA)}${
+              conflict.exampleB ? ` vs ${JSON.stringify(conflict.exampleB)}` : ""})` : ""
+            errors.push({
+              error: `Overlapping tokens ${term.name} and ${conflicting.name} used in same context${example}\n` +
+                `After: ${state.set[0].trail()}`,
+              conflict
+            })
+          }
+          addToSet(terms, term)
+          addToSet(incompatible, conflicting)
+        }
+      }
+
+      let tokenGroup = null
+      for (let group of groups) {
+        if (incompatible.some(term => group.tokens.includes(term))) continue
+        for (let term of terms) addToSet(group.tokens, term)
+        tokenGroup = group
+        break
+      }
+      if (!tokenGroup) {
+        tokenGroup = new TokenGroup(terms, groups.length + id)
+        groups.push(tokenGroup)
+      }
+      // state.tokenGroup = tokenGroup.groupID
+    }
+
+    if (errors.length)
+      this.b.raise(errors.map(e => e.error).join("\n\n"))
+    if (groups.length + id > 16)
+      this.b.raise(`Too many different token groups (${groups.length}) to represent them as a 16-bit bitfield`)
+
+    let precTable = this.buildPrecTable(softConflicts)
```

Comment:
Also (mostly) a copy-paste from `buildTokenGroups`. However, I don't know if the commented-out code on Line1824 is necessary. Uncommenting it causes tests to fail, but I'm not sure about its semantics soooo...